### PR TITLE
Order of glob & rglob return iterable is now deterministic

### LINKED ISSUE #140: Enforce order in which data is written / loaded to / from disk
Currently, the `DataFileCollector` relies on the order in which the filesystem returns filepaths to load the serialised data.
This leads to non-determinism, as demonstrated by failing CIs. 

In order to mitigate this, all the performance data for a singular test should be written into a singular file, where it can be reliably loaded from in the correct order (or something similar to this)

### PR DESCRIPTION
**Pull request by RecurvedBow**
Closes #140

### CODE REVIEW

File: fetching/strat.py [hunk 1]
```diff
@@ -23,8 +23,12 @@ def __init__(self, overwrite_tests: bool = True, recurse_into_subdirs: bool = Tr
 
     def apply(self, project: Project):
         assert project.test_directory is not None
+        
+        file_paths = self.globber(project.test_directory, "*")
+        # Ensures that the order is deterministic.
+        sorted_file_paths = sorted(file_paths)
         for path in filter(
-            self._is_test_file, self.globber(project.test_directory, "*")
+            self._is_test_file, sorted_file_paths
```
Comment (bengsparks) [inline fetching/strat.py:31]: Why does sorting need to be done here too?

File: tests/fetching/test_strat.py [hunk 2]
```diff
@@ -28,9 +28,13 @@ def test_if_test_object_searches_for_test_files_in_subfolders_it_generates_test_
...
     )
     test_object.apply(project_folder)
+
+    potential_test_file_paths = cwd.rglob("test_*.py")
+    # Ensures that the order is deterministic.
+    sorted_potential_test_file_paths = sorted(potential_test_file_paths)
     test_file_paths = list(
         filter(
-            lambda s: not s.name.endswith(PyTestStrategy.SUFFIX), cwd.rglob("test_*.py")
+            lambda s: not s.name.endswith(PyTestStrategy.SUFFIX), sorted_potential_test_file_paths
```
Comment (bengsparks) [inline tests/fetching/test_strat.py:37]: This too

File: typegen/evaluation/file_type_hints_collector.py [hunk 3]
```diff
@@ -33,7 +33,9 @@ def collect_data_from_folder(
         assert folder.is_dir(), f"{folder} is not a folder path."
         file_pattern = "*.py"
         file_paths = folder.rglob(file_pattern) if include_also_files_in_subdirectories else folder.glob(file_pattern)
-        self.collect_data(root, file_paths)
+        # Ensures that the order is deterministic.
+        sorted_file_paths = sorted(file_paths)
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:37]: Here too

File: typegen/trace_data_file_collector.py [hunk 4]
```diff
@@ -24,7 +24,9 @@ def collect_data(
             potential_trace_data_file_paths = path.rglob(self.file_pattern)
         else:
             potential_trace_data_file_paths = path.glob(self.file_pattern)
-        for potential_trace_data_file_path in potential_trace_data_file_paths:
+        # Ensures that the order is deterministic.
+        sorted_potential_trace_data_file_paths = sorted(potential_trace_data_file_paths)
```
Comment (bengsparks) [inline typegen/trace_data_file_collector.py:28]: This is the one we need. :+1:

**Review by bengsparks — CHANGES_REQUESTED** [review]

File: tests/fetching/test_strat.py (hunk 2 above)
Comment (RecurvedBow) [reply tests/fetching/test_strat.py]: Done

File: fetching/strat.py (hunk 1 above)
Comment (RecurvedBow) [reply fetching/strat.py]: Removed

File: tests/fetching/test_strat.py (hunk 2 above)
Comment (RecurvedBow) [reply tests/fetching/test_strat.py]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 3 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:37]: Is necessary

**Review by bengsparks — APPROVED** [review]

**Pull request merged**